### PR TITLE
Hugo/novel details view migration

### PR DESCRIPTION
**Pull request by zhanhugo**
Changed TextView for novel synopsis to be displayed with a DefinableTextView and wrapped the contents of NovelDetailsView with a DefinerView. Refactored and moved nongesture functions out of the coordinator and into a newly created DefinableTextViewModel. This fixes a bug where the Coordinator does not receive the latest tokensRange, which caused the tap gesture to not highlight words.
 
Screen Shot 2021-07-08 at 6 09 31 PM

### CODE REVIEW

File: Reed/Components/DefinableText/DefinableText.swift [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        
+        @objc func wordTapped(gesture: UITapGestureRecognizer) {
+            let textView = gesture.view as! DefinableTextView
+            let location = gesture.location(in: textView)
+            let position = CGPoint(x: location.x + textView.font.pointSize / 2, y: location.y)
+            let lineArray = CTFrameGetLines(textView.ctFrame!) as! Array<CTLine>
+            let tappedLine = viewModel.getLine(lineY: textView.linesYCoordinates!, l: 0, r: textView.linesYCoordinates!.count - 1, y: position.y)
+            if tappedLine > -1 {
+                let tappedIndex = CTLineGetStringIndexForPosition(lineArray[tappedLine], position) - 1
+                print(viewModel.tokensRange)
```
Comment (Rawgers) [inline Reed/Components/DefinableText/DefinableText.swift]: Remove

File: Reed/Components/DefinableText/DefinableText.swift [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    var definitions: [Definition]
+    
+    static func == (lhs: DefinitionDetails, rhs: DefinitionDetails) -> Bool {
+        return lhs.definitions == rhs.definitions
+    }
+}
+
+struct DefinableText: UIViewRepresentable {
+    @Binding var content: NSMutableAttributedString
+    @ObservedObject var viewModel: DefinableTextViewModel
```
Comment (Rawgers) [inline Reed/Components/DefinableText/DefinableText.swift]: Do we still need `@ObservedObject`? If so, should it be `@StateObject`?

File: Reed/Components/DefinableText/DefinableTextViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+//
+
+import Foundation
+import SwiftUI
+
+class DefinableTextViewModel: ObservableObject {
+    var tappedRange: NSRange!
+    var selectedRange: NSRange!
+    let dictionaryFetcher = DictionaryFetcher()
+    @Published var tokensRange: [Int]
```
Comment (Rawgers) [inline Reed/Components/DefinableText/DefinableTextViewModel.swift]: Can we get rid of `@Published` here?

File: Reed/Components/DefinableText/DefinableTextViewModel.swift [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    let definerResultHandler: ([DefinitionDetails]) -> Void
+    let hideNavHandler: () -> Void
+    let getTokenHandler: (Int, Int, Int) -> Token?
+    
+    internal init(tokensRange: [Int], definerResultHandler: @escaping ([DefinitionDetails]) -> Void, hideNavHandler: @escaping () -> Void, getTokenHandler: @escaping (Int, Int, Int) -> Token?) {
```
Comment (Rawgers) [inline Reed/Components/DefinableText/DefinableTextViewModel.swift]: Break up this function signature.

File: Reed/Components/DefinableText/DefinableTextViewModel.swift [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        }
+        self.definerResultHandler(entries)
+    }
+    
+    func highlightSelection(textView: DefinableTextView) {
+        if selectedRange != nil {
+            textView.content.addAttribute(NSAttributedString.Key.backgroundColor, value: UIColor.clear, range: selectedRange!)
+        }
+        selectedRange = tappedRange
+        textView.content.addAttribute(NSAttributedString.Key.backgroundColor, value: UIColor.systemYellow.withAlphaComponent(0.3), range: selectedRange!)
```
Comment (Rawgers) [inline Reed/Components/DefinableText/DefinableTextViewModel.swift:106]: Break up this call into separate lines.

File: Reed/NovelDetails/views/NovelDetailsView.swift [hunk 6]
```diff
@@ -62,23 +62,39 @@ struct NovelDetailsView: View {
...
                             .foregroundColor(Color(.systemGray4))
                             .padding(.bottom, 4)
-                        Text(viewModel.novelSynopsis)
+                        ZStack {
+                            if viewModel.isNovelSynopsisProcessing {
+                                ProgressView()
+                                    .frame(width: DefinerConstants.CONTENT_WIDTH, height: self.viewModel.novelSynopsisHeight)
+                            } else {
+                                Rectangle()
+                                    .frame(width: DefinerConstants.CONTENT_WIDTH, height: self.viewModel.novelSynopsisHeight)
```
Comment (Rawgers) [inline Reed/NovelDetails/views/NovelDetailsView.swift]: Break up call to `frame`.

**Review by Rawgers — CHANGES_REQUESTED** [review]

File: Reed/Components/DefinableText/DefinableText.swift (hunk 1 above)
Comment (zhanhugo) [reply Reed/Components/DefinableText/DefinableText.swift]: resolved

File: Reed/Components/DefinableText/DefinableText.swift (hunk 2 above)
Comment (zhanhugo) [reply Reed/Components/DefinableText/DefinableText.swift]: removed

File: Reed/Components/DefinableText/DefinableTextViewModel.swift (hunk 4 above)
Comment (zhanhugo) [reply Reed/Components/DefinableText/DefinableTextViewModel.swift]: fixed

File: Reed/Components/DefinableText/DefinableTextViewModel.swift (hunk 3 above)
Comment (zhanhugo) [reply Reed/Components/DefinableText/DefinableTextViewModel.swift]: removed

File: Reed/Components/DefinableText/DefinableTextViewModel.swift (hunk 5 above)
Comment (zhanhugo) [reply Reed/Components/DefinableText/DefinableTextViewModel.swift:106]: fixed

File: Reed/NovelDetails/views/NovelDetailsView.swift (hunk 6 above)
Comment (zhanhugo) [reply Reed/NovelDetails/views/NovelDetailsView.swift]: fixed

File: Reed/Components/DefinableText/DefinableTextViewModel.swift (hunk 5 above)
Comment (Rawgers) [reply Reed/Components/DefinableText/DefinableTextViewModel.swift:106]: Did you make sure to add stage and commit this change?

File: Reed/Components/DefinableText/DefinableTextViewModel.swift [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+                var specicificLexemes = ""
+                if !sense.specificLexemes.isEmpty {
+                    for specific in sense.specificLexemes {
+                        specicificLexemes += specific + ", "
+                    }
+                    specicificLexemes.removeLast(2)
+                }
+                definitions.append(Definition(specicificLexemes: specicificLexemes, definition: definition))
+            }
+            entries.append(DefinitionDetails(title: tappedWord, primaryReading: primaryReading, terms: terms, definitions: definitions))
```
Comment (Rawgers) [inline Reed/Components/DefinableText/DefinableTextViewModel.swift]: Break this up

File: Reed/Components/DefinableText/DefinableText.swift [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+        
+        @objc func wordTapped(gesture: UITapGestureRecognizer) {
+            let textView = gesture.view as! DefinableTextView
+            let location = gesture.location(in: textView)
+            let position = CGPoint(x: location.x + textView.font.pointSize / 2, y: location.y)
+            let lineArray = CTFrameGetLines(textView.ctFrame!) as! Array<CTLine>
+            let tappedLine = viewModel.getLine(lineY: textView.linesYCoordinates!, l: 0, r: textView.linesYCoordinates!.count - 1, y: position.y)
+            if tappedLine > -1 {
+                let tappedIndex = CTLineGetStringIndexForPosition(lineArray[tappedLine], position) - 1
+                if let token = viewModel.getTokenHandler(viewModel.tokensRange[1], viewModel.tokensRange[2], tappedIndex + viewModel.tokensRange[0]) {
```
Comment (Rawgers) [inline Reed/Components/DefinableText/DefinableText.swift]: Break this up.

**Review by Rawgers — CHANGES_REQUESTED** [review]
Touch up the remaining nits where the lines are too long, and make sure to add a screenshot and description to the PR.

File: Reed/Components/DefinableText/DefinableTextViewModel.swift (hunk 7 above)
Comment (zhanhugo) [reply Reed/Components/DefinableText/DefinableTextViewModel.swift]: fixed

File: Reed/Components/DefinableText/DefinableText.swift (hunk 8 above)
Comment (zhanhugo) [reply Reed/Components/DefinableText/DefinableText.swift]: fixed

**Comment by Rawgers** [comment]
LGTM.
Having trouble pulling your branch, so I'm going to merge and pull from master.

**Review by Rawgers — APPROVED** [review]

**Pull request merged**